### PR TITLE
feat(Forwarded-header): add configurable x-forwarded-* header to routes and services

### LINKED ISSUE #5559: Configurable X-Forwarded-* headers on services and routes
### Summary

Kong automatically assign X-Forwarded-* headers based on some magic (in my case definitely wrong logic). So it would be great if I could overwrite those headers based on routes and services like I would do in nginx location directives. 

### Solution

Add optional properties for x-forwarded-* header to services and routes. If those fields are empty use the magic to define x-forwarded headers, otherwise take the values

### PR DESCRIPTION
**Pull request by trollr**
### Summary

Add X-Forwarded-* to routes and services

### Full changelog

Add following configurable headers to services and routes:

- X-Forwarded-Proto
- X-Forwarded-Host
- X-Forwarded-Port

### Issues resolved

Fix #5559, #5018, #3461

### CODE REVIEW

**Comment by bungle** [comment]
See my comment here:
https://github.com/Kong/kong/issues/5559#issuecomment-585380290

File: kong/runloop/handler.lua [hunk 1]
```diff
@@ -1041,14 +1041,14 @@ return {
 
       local trusted_ip = kong.ip.is_trusted(realip_remote_addr)
       if trusted_ip then
-        forwarded_proto = var.http_x_forwarded_proto or scheme
-        forwarded_host  = var.http_x_forwarded_host  or host
-        forwarded_port  = var.http_x_forwarded_port  or port
+        forwarded_proto = route.x_forwarded_proto or service.x_forwarded_proto or var.http_x_forwarded_proto or scheme
```
Comment (bungle) [inline kong/runloop/handler.lua]: In general these (trusted) ones should be in this order:

```suggestion
        forwarded_proto = var.http_x_forwarded_proto or route.x_forwarded_proto or service.x_forwarded_proto or scheme
```

But I feel like we use a sledgehammer to implement this. At least we should only have this on route or service but not on both. Perhaps only on service. Not sure if this should be on `db` or say ENV variable that you pass to Kong when it is started.

Comment (bungle) [reply kong/runloop/handler.lua]: This is how Amazon does it, and this works just fine with Kong:
https://kevinhakanson.com/2018-03-15-amazon-load-balancers-x-forwarded-headers-and-proxy-protocol-support

I am not sure why Docker cannot do it.

**Comment by bungle** [comment]
Perhaps something like:
```
KONG_PORTMAP="80:8000, 443:8443" kong start
```

File: kong/runloop/handler.lua (hunk 1 above)
Comment (trollr) [reply kong/runloop/handler.lua]: If we use a env var for this case I see two issues:

1. It's a implicit behaviour for every registered service and will cause new issues. 
2. Maybe I want to set different values for the services. 

In my opinion my solution proposal is the cleanest possible way to allow people to configure kong as they need it

**Comment by trollr** [comment]
> Perhaps something like:
> 
> ```
> KONG_PORTMAP="80:8000, 443:8443" kong start
> ```
> 
> That is used to change use of `var.server_port` to matching mapped one, for `X-Forwarded-Port`. Do you have issues with other headers too?

That's exactly what happens in the k8s cluster. However it use the internal docker port to set x-forwarded-port

**Comment by trollr** [comment]
Removed the x-forwarded-* header configuration for routes

File: kong/runloop/handler.lua (hunk 1 above)
Comment (bungle) [reply kong/runloop/handler.lua]: Ah, ok I see, you want to break X-Forwarded even in case where  you trust them.

**Comment by bungle** [comment]
Before we go as deep as integrating something very specific to our core entities, like `X-Forwarded-*` headers. We should slice this to actual issue, you are having,

1. Do you have problems with other `X-Forwarded-*`  headers than `X-Forwarded-Port`, and can you describe the issue?

2. Right now this can be fixed by a plugin. And I think it is preferred to changing the core entities.

If it is only the `Port` that is causing a problem, perhaps we can add a more generic thing in the core entity. But then again 2. will solve this the way you want it to be solved.

**Comment by trollr** [comment]
> Before we go as deep as integrating something very specific to our core entities, like `X-Forwarded-*` headers. We should slice this to actual issue, you are having,
> 
> 1. Do you have problems with other `X-Forwarded-*`  headers than `X-Forwarded-Port`, and can you describe the issue?
> 2. Right now this can be fixed by a plugin. And I think it is preferred to changing the core entities.
> 
> If it is only the `Port` that is causing a problem, perhaps we can add a more generic thing in the core entity. But then again 2. will solve this the way you want it to be solved.

At the moment I only have problems with the x-forwarded-port header described here: #5559

Could you please give me a hint how to fix this issue with a plugin? The request transformer plugin wont work to modify the x-forwarded-* headers. If there is a clear way how to do this I could contribute it or change the request transformer plugin to support it.

**Comment by bungle** [comment]
@trollr,

Here is my quick and dirty `Forwarded` plugin that allows you to do what this PR does, but by plugin: https://github.com/Kong/kong-plugin-forwarded

Let me know what you think about it, and does it work for you?

**Comment by bungle** [comment]
If you want you can just do that what I did there in a plugin in a serverless function directly.

**Comment by bungle** [comment]
Here is an example how you can do it with `post-function`:
```
http :8001/services/bin/plugins \
  name=post-function \
 config.functions="\
ngx.var.upstream_x_forwarded_host='test.host';\
ngx.var.upstream_x_forwarded_port=1234;\
ngx.var.upstream_x_forwarded_proto='http'" -f
```

**Comment by trollr** [comment]
Oh nice. That was easy. 

Solution for kong ingress controller
```
apiVersion: configuration.konghq.com/v1
kind: KongPlugin
metadata:
  name: post-function
  namespace: fusionauth
config:
  functions:
    - "ngx.var.upstream_x_forwarded_port=443; -f"
plugin: serverless-functions
```

Thanks @bungle for your support :) Happy now

**Comment by bungle** [comment]
@trollr,

Thank you too! Let us consider if we want to add such a plugin to core.

Btw. one typo on your config:

```
- "ngx.var.upstream_x_forwarded_port=443; -f"
```
should be just:
```
- ngx.var.upstream_x_forwarded_port=443
```

**Comment by bungle** [comment]
And if you want to set more than one, it is quite easy as well:

```yaml
apiVersion: configuration.konghq.com/v1
kind: KongPlugin
metadata:
  name: post-function
  namespace: fusionauth
config:
  functions:
    - ngx.var.upstream_x_forwarded_host=myhost.test
    - ngx.var.upstream_x_forwarded_port=443
    - ngx.var.upstream_x_forwarded_proto=https
plugin: serverless-functions
```

**Pull request closed without merging**